Homework2-7/Server/src/main/java/Server.java

```diff
@@ -0,0 +1,80 @@
+import java.io.IOException;
+import java.net.ServerSocket;
+import java.net.Socket;
+import java.util.concurrent.ConcurrentLinkedDeque;
+
+public class Server {
+
+    private static final int DEFAULT_PORT = 8189;
+
+    private ConcurrentLinkedDeque<ClientHandler> clients;
+
+    public Server(int port) {
+        clients = new ConcurrentLinkedDeque<>();
+        try (ServerSocket server = new ServerSocket(port)) {
+            System.out.println("[DEBUG] server started on port: " + port);
+            while (true) {
+                Socket socket = server.accept(); // get connection
+                System.out.println("[DEBUG] client accepted");
+                ClientHandler handler = new ClientHandler(socket, this);
+                addClient(handler);
+                new Thread(handler).start();
+            }
+        } catch (Exception e) {
+            System.err.println("Server was broken");
+        }
+    }
+
+    public void addClient(ClientHandler clientHandler){
+        clients.add(clientHandler);
+        System.out.println("[DEBUG] client added to broadcast queue");
+    }
+
+    public void removeClient(ClientHandler clientHandler){
+        clients.remove(clientHandler);
+        System.out.println("[DEBUG] client removed from broadcast queue");
+    }
+
+    public void broadCastMessage(String msg) throws IOException {
+        for (ClientHandler client : clients) {
+            client.sendMessage(msg);
+        }
+    }
+
+    public void sendMessageTo(String nickname, String msg) throws IOException {
+        String nick = cutNick(msg);
+        for (ClientHandler recipient : clients){
```

Comment:
if(!clients.contains(nick)) {
throw new IllegalArgumentException("User with nick ? not found");
}